astraguard/swarm/leader_election.py

```
@@ -0,0 +1,343 @@
+                    )
+        quorum_size = self._calculate_quorum_size()
+        if len(self.votes_received) >= quorum_size:
+            logger.info(f"{self.config.agent_id.satellite_serial} achieved quorum ({len(self.votes_received)}/{quorum_size}), becoming leader")
+            elapsed_ms = (datetime.now() - self.election_start_time).total_seconds() * 1000
+            self.metrics.convergence_time_ms = elapsed_ms
+            self.metrics.election_count += 1
+            await self._become_leader()
+            return
+        if self.election_start_time:
+            election_duration = (datetime.now() - self.election_start_time).total_seconds() * 1000
+            if election_duration > randint(self.ELECTION_TIMEOUT_MIN_MS, self.ELECTION_TIMEOUT_MAX_MS):
```

Ruff (0.14.10)

Standard pseudo-random generators are not suitable for cryptographic purposes

(S311)